pkg/controller/cronjob/cronjob_controllerv2.go

```diff
@@ -0,0 +1,673 @@
+				return nil, err
+			}
+			if !deleteJob(cj, job, jc, recorder) {
+				return nil, fmt.Errorf("could not replace job %s/%s", job.Namespace, job.Name)
+			}
+		}
+	}
+
+	jobReq, err := getJobFromTemplate(cj, scheduledTime)
+	if err != nil {
+		klog.ErrorS(err, "Unable to make Job from template", "cronjob", klog.KRef(cj.GetNamespace(), cj.GetName()))
+		return nil, err
+	}
+	jobResp, err := jc.CreateJob(cj.Namespace, jobReq)
+	switch {
+	case errors.HasStatusCause(err, v1.NamespaceTerminatingCause):
+		// TODO: @alpatel log, event?
+	case errors.IsAlreadyExists(err):
+		// TODO: @alpatel handle this, we tried to create a job that already exists. may be update/patch?
+	case err != nil:
+		// default error handling
+		recorder.Eventf(cj, v1.EventTypeWarning, "FailedCreate", "Error creating job: %v", err)
+		return nil, err
+	}
+	klog.V(4).InfoS("Created Job", "job", klog.KRef(jobResp.GetNamespace(), jobResp.GetName()), "cronjob", klog.KRef(cj.GetNamespace(), cj.GetName()))
+	recorder.Eventf(cj, v1.EventTypeNormal, "SuccessfulCreate", "Created job %v", jobResp.Name)
+
+	// ------------------------------------------------------------------ //
+
+	// If this process restarts at this point (after posting a job, but
+	// before updating the status), then we might try to start the job on
+	// the next time.  Actually, if we re-list the SJs and Jobs on the next
+	// iteration of syncAll, we might not see our own status update, and
+	// then post one again.  So, we need to use the job name as a lock to
+	// prevent us from making the job twice (name the job with hash of its
+	// scheduled time).
+
+	// Add the just-started job to the status list.
+	jobRef, err := getRef(jobResp)
+	if err != nil {
+		klog.V(2).InfoS("Unable to make object reference", "cronjob", klog.KRef(cj.GetNamespace(), cj.GetName()), "err", err)
+		return nil, fmt.Errorf("unable to make object reference for job for %s", klog.KRef(cj.GetNamespace(), cj.GetName()))
+	}
+	cj.Status.Active = append(cj.Status.Active, *jobRef)
+	cj.Status.LastScheduleTime = &metav1.Time{Time: scheduledTime}
+	if _, err := cjc.UpdateStatus(cj); err != nil {
+		klog.InfoS("Unable to update status", "cronjob", klog.KRef(cj.GetNamespace(), cj.GetName()), "resourceVersion", cj.ResourceVersion, "err", err)
+		return nil, fmt.Errorf("unable to update status for %s (rv = %s): %v", klog.KRef(cj.GetNamespace(), cj.GetName()), cj.ResourceVersion, err)
+	}
+
+	t := nextScheduledTimeDurationWithDelta(sched, now)
+	return t, nil
+}
+
+// nextScheduledTimeDurationWithDelta returns the time duration to requeue based on
+// the schedule and current time. It adds a 100ms padding to the next requeue to account
+// for Network Time Protocol(NTP) time skews. If the time drifts are adjusted which in most
+// realistic cases would be around 100s, scheduled cron will still be executed without missing
+// the schedule.
+func nextScheduledTimeDurationWithDelta(sched cron.Schedule, now time.Time) *time.Duration {
```

Comment:
just `nextScheduledTimeDuration` with delta is implementation detail.

Reply:
updated